File: src/core/analyzer.ts
import type {
  Detector,
  DetectionSignal,
  LineRange,
  ScanResult,
  ScanSummary,
  AnalysisMetadata,
  FileMetadata,
  ParsedCode,
} from '../types.js';
import { selectParser, isBinaryContent } from '../parsers/parser.js';
import { classifyWindows, mergeWindows } from './segmenter.js';
import { overallConfidenceLevel } from './confidence.js';
import { entropyDetector } from '../detectors/entropy.js';
import { commentPatternsDetector } from '../detectors/comment-patterns.js';
import { namingPatternsDetector } from '../detectors/naming-patterns.js';
import { structuralDetector } from '../detectors/structural.js';
import {
  modelSignaturesDetector,
  scoreClaudePatterns,
  scoreGptPatterns,
  scoreCopilotPatterns,
  attributeModel,
} from '../detectors/model-signatures.js';
import { readFileSync } from 'node:fs';
import { resolve, relative } from 'node:path';

// Package version — read once at startup
let toolVersion = '0.1.0';
try {
  const pkg = JSON.parse(readFileSync(new URL('../../package.json', import.meta.url), 'utf-8'));
  toolVersion = pkg.version ?? '0.1.0';
} catch {
  // fallback
}

/**
 * All registered detectors. New detectors are added here.
 */
const detectors: Detector[] = [
  entropyDetector,
  commentPatternsDetector,
  namingPatternsDetector,
  structuralDetector,
  modelSignaturesDetector,
];

/**
 * Register an additional detector (e.g., model-signatures, ngram).
 */
export function registerDetector(detector: Detector): void {
  detectors.push(detector);
}

/**
 * Analyze a single file for AI-generated code.
 */
export function analyze(filePath: string): ScanResult {
  const startTime = performance.now();
  const absPath = resolve(filePath);
  const relPath = relative(process.cwd(), absPath);

  // Read file
  let content: string;
  try {
    content = readFileSync(absPath, 'utf-8');
  } catch (err) {
    throw new Error(`Cannot read file: ${absPath} — ${(err as Error).message}`);
  }

  // Binary check
  if (isBinaryContent(content)) {
    throw new Error(`Binary file not supported: ${relPath}`);
  }

  // Empty file
  const lines = content.split('\n');
  if (lines.length === 0 || content.trim() === '') {
    return emptyResult(absPath, relPath, startTime);
  }

  // Parse
  const parser = selectParser(filePath);
  const parsed: ParsedCode = parser.parse(content, filePath);

  // Run all detectors
  const allSignals: DetectionSignal[] = [];
  for (const detector of detectors) {
    const signals = detector.detect(parsed);
    allSignals.push(...signals);
  }

  // Segment into windows and merge into line ranges
  const windows = classifyWindows(parsed.lines, allSignals);
  const rawRanges = mergeWindows(windows, lines.length);

  // Attach model attribution to AI-generated ranges
  const modelScores = [
    scoreClaudePatterns(parsed),
    scoreGptPatterns(parsed),
    scoreCopilotPatterns(parsed),
  ];
  const attribution = attributeModel(modelScores);

  const ranges = rawRanges.map((range) => {
    if (range.classification === 'ai-generated' && attribution) {
      return { ...range, modelAttribution: attribution };
    }
    return range;
  });

  // Compute summary
  const summary = computeSummary(ranges, lines.length);

  // Build metadata
  const duration = Math.round(performance.now() - startTime);
  const metadata = buildMetadata(duration);

  const file: FileMetadata = {
    path: absPath,
    relativePath: relPath,
    totalLines: lines.length,
    language: parsed.language,
    parserUsed: parser.language,
  };

  return { file, ranges, summary, metadata };
}

function computeSummary(ranges: readonly LineRange[], totalLines: number): ScanSummary {
  let aiLines = 0;
  let humanLines = 0;
  let unknownLines = 0;
  const modelCounts: Record<string, number> = {};

  for (const range of ranges) {
    const rangeLines = range.endLine - range.startLine + 1;
    switch (range.classification) {
      case 'ai-generated':
        aiLines += rangeLines;
        break;
      case 'human-written':
        humanLines += rangeLines;
        break;
      default:
        unknownLines += rangeLines;
    }

    if (range.modelAttribution) {
      const model = range.modelAttribution.model;
      modelCounts[model] = (modelCounts[model] ?? 0) + rangeLines;
    }
  }

  const total = aiLines + humanLines + unknownLines;
  return {
    aiPercentage: total > 0 ? Math.round((aiLines / total) * 100) : 0,
    humanPercentage: total > 0 ? Math.round((humanLines / total) * 100) : 0,
    unknownPercentage: total > 0 ? Math.round((unknownLines / total) * 100) : 0,
    overallConfidence: overallConfidenceLevel(ranges, totalLines),
    modelBreakdown: modelCounts,
  };
}

function buildMetadata(duration: number): AnalysisMetadata {
  return {
    toolVersion,
    analyzedAt: new Date().toISOString(),
    algorithmVersions: Object.fromEntries(
      detectors.map((d) => [d.name, d.version]),
    ),
    thresholds: {
      entropyAiCeiling: 4.0,
      entropyHumanFloor: 4.5,
      aiScoreThreshold: 0.60,
      humanScoreThreshold: 0.40,
      sigmoidK: 10,
      sigmoidMidpoint: 0.5,
    },
    duration,
  };
}

function emptyResult(absPath: string, relPath: string, startTime: number): ScanResult {
  return {
    file: {
      path: absPath,
      relativePath: relPath,
      totalLines: 0,
      language: 'unknown',
      parserUsed: 'none',
    },
    ranges: [],
    summary: {
      aiPercentage: 0,
      humanPercentage: 0,
      unknownPercentage: 0,
      overallConfidence: 'LOW',
      modelBreakdown: {},
    },
    metadata: buildMetadata(Math.round(performance.now() - startTime)),
  };
}

File: code_provenance/core/analyzer.py
"""Core analysis engine for Code Provenance."""

from __future__ import annotations

import time
from datetime import datetime, timezone
from pathlib import Path

from code_provenance import __version__
from code_provenance.types import (
    AnalysisMetadata,
    Classification,
    ConfidenceLevel,
    DetectionSignal,
    FileMetadata,
    LineRange,
    ParsedCode,
    ScanResult,
    ScanSummary,
)
from code_provenance.parsers.parser import select_parser, is_binary_content
from code_provenance.core.segmenter import classify_windows, merge_windows
from code_provenance.core.confidence import overall_confidence_level
from code_provenance.detectors.entropy import entropy_detector
from code_provenance.detectors.comment_patterns import comment_patterns_detector
from code_provenance.detectors.naming_patterns import naming_patterns_detector
from code_provenance.detectors.structural import structural_detector
from code_provenance.detectors.model_signatures import (
    model_signatures_detector,
    score_claude_patterns,
    score_gpt_patterns,
    score_copilot_patterns,
    attribute_model,
)

# All registered detectors. New detectors are added here.
_detectors: list = [
    entropy_detector,
    comment_patterns_detector,
    naming_patterns_detector,
    structural_detector,
    model_signatures_detector,
]


def register_detector(detector) -> None:
    """Register an additional detector (e.g., model-signatures, ngram)."""
    _detectors.append(detector)


def analyze(file_path: str) -> ScanResult:
    """Analyze a single file for AI-generated code."""
    start_time = time.perf_counter()
    abs_path = str(Path(file_path).resolve())
    try:
        rel_path = str(Path(abs_path).relative_to(Path.cwd()))
    except ValueError:
        rel_path = abs_path

    # Read file
    try:
        content = Path(abs_path).read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError) as err:
        raise RuntimeError(f"Cannot read file: {abs_path} -- {err}") from err

    # Binary check
    if is_binary_content(content):
        raise RuntimeError(f"Binary file not supported: {rel_path}")

    # Empty file
    lines = content.split("\n")
    if not lines or content.strip() == "":
        return _empty_result(abs_path, rel_path, start_time)

    # Parse
    parser = select_parser(file_path)
    parsed: ParsedCode = parser.parse(content, file_path)

    # Run all detectors
    all_signals: list[DetectionSignal] = []
    for detector in _detectors:
        signals = detector.detect(parsed)
        all_signals.extend(signals)

    # Segment into windows and merge into line ranges
    windows = classify_windows(parsed.lines, all_signals)
    raw_ranges = merge_windows(windows, len(lines))

    # Attach model attribution to AI-generated ranges
    model_scores = [
        score_claude_patterns(parsed),
        score_gpt_patterns(parsed),
        score_copilot_patterns(parsed),
    ]
    attribution = attribute_model(model_scores)

    ranges: list[LineRange] = []
    for r in raw_ranges:
        if r.classification == Classification.AI_GENERATED and attribution:
            ranges.append(LineRange(
                start_line=r.start_line,
                end_line=r.end_line,
                classification=r.classification,
                confidence=r.confidence,
                model_attribution=attribution,
                signals=r.signals,
            ))
        else:
            ranges.append(r)

    # Compute summary
    summary = _compute_summary(ranges, len(lines))

    # Build metadata
    duration = round((time.perf_counter() - start_time) * 1000)
    metadata = _build_metadata(duration)

    file_meta = FileMetadata(
        path=abs_path,
        relative_path=rel_path,
        total_lines=len(lines),
        language=parsed.language,
        parser_used=parser.language,
    )

    return ScanResult(
        file=file_meta,
        ranges=tuple(ranges),
        summary=summary,
        metadata=metadata,
    )


def _compute_summary(ranges: tuple[LineRange, ...] | list[LineRange], total_lines: int) -> ScanSummary:
    """Compute scan summary from line ranges."""
    ai_lines = 0
    human_lines = 0
    unknown_lines = 0
    model_counts: dict[str, int] = {}

    for r in ranges:
        range_lines = r.end_line - r.start_line + 1
        if r.classification == Classification.AI_GENERATED:
            ai_lines += range_lines
        elif r.classification == Classification.HUMAN_WRITTEN:
            human_lines += range_lines
        else:
            unknown_lines += range_lines

        if r.model_attribution:
            model = r.model_attribution.model.value
            model_counts[model] = model_counts.get(model, 0) + range_lines

    total = ai_lines + human_lines + unknown_lines

    # Build range dicts for confidence calculation
    range_dicts = [
        {
            "confidence": r.confidence,
            "classification": r.classification.value,
            "start_line": r.start_line,
            "end_line": r.end_line,
        }
        for r in ranges
    ]

    return ScanSummary(
        ai_percentage=round(ai_lines / total * 100) if total > 0 else 0,
        human_percentage=round(human_lines / total * 100) if total > 0 else 0,
        unknown_percentage=round(unknown_lines / total * 100) if total > 0 else 0,
        overall_confidence=overall_confidence_level(range_dicts, total_lines),
        model_breakdown=model_counts,
    )


def _build_metadata(duration: int) -> AnalysisMetadata:
    """Build analysis metadata."""
    return AnalysisMetadata(
        tool_version=__version__,
        analyzed_at=datetime.now(timezone.utc).isoformat(),
        algorithm_versions={d.name: d.version for d in _detectors},
        thresholds={
            "entropyAiCeiling": 4.0,
            "entropyHumanFloor": 4.5,
            "aiScoreThreshold": 0.60,
            "humanScoreThreshold": 0.40,
            "sigmoidK": 10,
            "sigmoidMidpoint": 0.5,
        },
        duration=duration,
    )


def _empty_result(abs_path: str, rel_path: str, start_time: float) -> ScanResult:
    """Return result for empty files."""
    duration = round((time.perf_counter() - start_time) * 1000)
    return ScanResult(
        file=FileMetadata(
            path=abs_path,
            relative_path=rel_path,
            total_lines=0,
            language="unknown",
            parser_used="none",
        ),
        ranges=(),
        summary=ScanSummary(
            ai_percentage=0,
            human_percentage=0,
            unknown_percentage=0,
            overall_confidence=ConfidenceLevel.LOW,
            model_breakdown={},
        ),
        metadata=_build_metadata(duration),
    )
